### PR TITLE
Toast

### PR DESCRIPTION
**Pull request by FarhadA97**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,8 @@
 node_modules
 /.pnp
 .pnp.js
+dist/**/*
```
Comment (13bscsdahmed) [inline .gitignore]: just add dist and do rm -ef node_modules and then commit,. should work

File: client/.gitignore [hunk 2]
```diff
@@ -4,6 +4,7 @@
 /node_modules
 /.pnp
 .pnp.js
+package-lock.json
```
Comment (13bscsdahmed) [inline client/.gitignore]: this should not be in gitignore, research about the reason for doing so

File: client/src/components/Auth/AuthForm.module.css [hunk 3]
```diff
@@ -58,6 +58,14 @@
...
     border-color: #873abb;
   }
+
+.actions button:disabled,
+.actions button:disabled:hover,
+.actions button:disabled:active {
+  background-color: #ccc;
+  color: #656262;
+  border-color: #ccc;
+}
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.module.css]: please make separate configurable components for commonly used elements like inputs, button and move the relevant styles to the component's styles file

Comment (13bscsdahmed) [reply client/src/components/Auth/AuthForm.module.css]: please do this for all elements in a separate commit

File: client/src/components/Auth/AuthForm.tsx [hunk 4]
```diff
@@ -1,121 +1,184 @@
...
+  }
+
+  let toastProperties: IToastProps ;
+
+  const showToastHandler : (type:string, message:string) => void = (type,message) =>{
+    switch(type) {
+      case 'success':
+        toastProperties = {
+          id: list.length+1,
+          title: 'Success',
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: title should be an enum

File: client/src/components/Auth/AuthForm.tsx [hunk 5]
```diff
@@ -1,121 +1,184 @@
...
+          description: message,
+          backgroundColor: '#5cb85c'
+        }
+        break;
+      case 'fail':
+        toastProperties = {
+          id: list.length+1,
+          title: 'Error',
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: same,. title should be an enum

File: client/src/components/Auth/AuthForm.tsx [hunk 6]
```diff
@@ -1,121 +1,184 @@
...
-    catch (error) 
-    {
-        setIsLoading(false);
-        const errObj = error as AxiosError
-        const err = errObj.response?.data;
-        const response = err as AxiosError;
-        const notify = () => toast.error(response.message);
-        notify();
+    try {
+      const response = await Axios({
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: for data fetching make separate services for each resource like users, todos etc and call the service methods in components

File: client/src/components/Auth/AuthForm.tsx [hunk 7]
```diff
@@ -1,121 +1,184 @@
...
+    if(!formIsValid){
+      return;
+    }
 
     setIsLoading(true);
     let url;
     if (isLogin) {
-      url =
-        'http://localhost:9000/user/login';
+      url = "http://localhost:9000/user/login";
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: dont use hardcoded urls

File: client/src/components/Auth/AuthForm.tsx [hunk 8]
```diff
@@ -1,121 +1,184 @@
...
     } else {
-      url =
-        'http://localhost:9000/user/register';
+      url = "http://localhost:9000/user/register";
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: same

File: client/src/components/Auth/AuthForm.tsx [hunk 9]
```diff
@@ -1,121 +1,184 @@
...
+          email,
+          password,
+        },
+      });
 
+      if (response.status === 200) {
+        // Could also use history.push('/login')
+        setIsLoading(false);
+        dispatch(authActions.login(response.data));
+        showToastHandler('success','login Successful')
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: should accept an object instead of passing args

File: client/src/components/Auth/AuthForm.tsx [hunk 10]
```diff
@@ -1,121 +1,184 @@
...
+        setTimeout(() => {
+          navigate("/");
+        }, 900);
+      } else {
+        setIsLoading(false);
+        console.log("unspecified");
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: remove log

File: client/src/components/Auth/AuthForm.tsx [hunk 11]
```diff
@@ -1,121 +1,184 @@
...
-}
+
+    nameResetHandler();
+    emailResetHandler();
+    passwordResetHandler();
+  };
+
+  const nameInputClasses = nameHasError ? `${classes["form-control"]} ${classes['invalid']}` : `${classes['form-control']}`;
+  const emailInputClasses = emailHasError ? `${classes["form-control"]} ${classes['invalid']}` : `${classes['form-control']}`;
+  const passwordInputClasses = passwordHasError ? `${classes["form-control"]} ${classes['invalid']}` : `${classes['form-control']}`;
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: should not be like this,. encapsulating inputs in separate component should fix this

File: client/src/components/Auth/AuthForm.tsx [hunk 12]
```diff
@@ -1,121 +1,184 @@
...
 
   return (
+    <>
+    <Toast toastList={list} position='bottom-right' setList={setList} />
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: why is the toast component here,.??? it should be in app.tsc and should be triggerable from anywhere inside the app

File: client/src/components/UI/Toast.tsx [hunk 13]
```diff
@@ -0,0 +1,58 @@
+import { totalmem } from "os";
```
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: what is this used for?

File: server/.env.example [hunk 14]
```diff
@@ -0,0 +1,3 @@
+PORT=9000
```
Comment (13bscsdahmed) [inline server/.env.example]: why the file is .env.example,. plus env files should not be part of the commit and should be in gitignore

File: client/src/components/UI/Toast.tsx (hunk 13 above)
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: why importing this

**Review by 13bscsdahmed — CHANGES_REQUESTED** [review]

File: client/src/components/UI/Toast.tsx (hunk 13 above)
Comment (FarhadA97) [reply client/src/components/UI/Toast.tsx]: It was an auto import by IDE. I removed it but didn't commit it then. It's removed now.

File: server/.env.example (hunk 14 above)
Comment (FarhadA97) [reply server/.env.example]: I've seen this file being added in few tutorials. Basically .env.example works as a guide to setup the actual .env. It has all the constant variables same as .env but with no actual values but only examples for what kind of value these variables should have. 
env file is already in gitignore.

File: client/src/components/Auth/AuthForm.tsx (hunk 12 above)
Comment (FarhadA97) [reply client/src/components/Auth/AuthForm.tsx]: Toast is now available app wide, also updated logic for showing toast. I've created a slice for toast and it can be triggered by dispatching showToast reducer function.

File: client/src/components/Auth/AuthForm.tsx (hunk 6 above)
Comment (FarhadA97) [reply client/src/components/Auth/AuthForm.tsx]: I've included a fetch custom hook for user sign in/sign up and then i've used the function returned by custom hook in AuthForm component.

File: client/src/components/Auth/AuthForm.module.css (hunk 3 above)
Comment (FarhadA97) [reply client/src/components/Auth/AuthForm.module.css]: I've made custom component for all the inputs in a separate commit.. Having a little issue with button components, i'll update it after discussing with you.

File: client/src/components/Auth/AuthForm.tsx (hunk 11 above)
Comment (FarhadA97) [reply client/src/components/Auth/AuthForm.tsx]: This is in seperate input component now.

File: client/package.json [hunk 15]
```diff
@@ -8,11 +8,11 @@
     "@testing-library/react": "^13.3.0",
     "@testing-library/user-event": "^13.5.0",
     "@types/jest": "^27.5.1",
-    "@types/node": "^16.11.36",
     "@types/react": "^18.0.9",
     "@types/react-dom": "^18.0.5",
     "@types/react-router-dom": "^5.3.3",
```
Comment (13bscsdahmed) [inline client/package.json]: all @types packages should be dev dependencies

File: .gitignore [hunk 16]
```diff
@@ -13,6 +15,7 @@ node_modules
 
 # misc
 .DS_Store
+.env
```
Comment (13bscsdahmed) [inline .gitignore:18]: you can add .env.* and it will ignore all env files,. .* is a wildcard operator and .env.* will mean add all files .starting with .env. to gitignore

File: client/src/components/Auth/AuthForm.tsx [hunk 17]
```diff
@@ -1,121 +1,163 @@
...
     setIsLogin((prevState) => !prevState);
   };
 
+  let formIsValid = false;
+  if (isLogin) {
+    if (emailIsValid && passwordIsValid) formIsValid = true;
+  } else {
+    if (nameIsValid && emailIsValid && passwordIsValid) formIsValid = true;
+  }
+
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx:51]: i would suggest exploring yup and formik npm packages later for forms and validations to make this code a lot more cleaner

File: client/src/components/Auth/AuthForm.tsx [hunk 18]
```diff
@@ -1,121 +1,163 @@
...
-        const errObj = error as AxiosError
-        const err = errObj.response?.data;
-        const response = err as AxiosError;
-        const notify = () => toast.error(response.message);
-        notify();
+    signIn({
+      method: "POST",
+      url: url,
+      data: {
+        name,
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: we shouldnt need name for signing in,. no?

Comment (13bscsdahmed) [reply client/src/components/Auth/AuthForm.tsx]: I think sign in should be a callable action in redux and we make the api call from the async thunk

File: client/src/components/Auth/AuthForm.tsx [hunk 19]
```diff
@@ -1,121 +1,163 @@
...
+      if(data.user.name.trim() !== ''){
+        dispatch(authActions.login(data));
+        dispatch(
+            toastActions.showToast({
+              type: Title.SUCCESS,
+              message: "Login Successful",
+            })
+          );
+      }
+      navigate('/');
```
Comment (13bscsdahmed) [inline client/src/components/Auth/AuthForm.tsx]: use navigate.push so that the history is maintained

File: client/src/components/Layout/MainNavigation.tsx [hunk 20]
```diff
@@ -14,8 +14,7 @@ const MainNavigation = () => {
   const logoutHandler = () => {
      dispatch(authActions.logout());
      navigate('/login');
-     navigate(0);
-    // optional: redirect the user
+
```
Comment (13bscsdahmed) [inline client/src/components/Layout/MainNavigation.tsx]: use navigate.push

File: client/src/components/UI/Button.tsx [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+const Button: React.FC<ButtonProps> = ({
+  children,
+  type,
+  classText,
+  disabled,
+  onClick,
+}) => {
+    const btnClass = `${classes[classText]}`
```
Comment (13bscsdahmed) [inline client/src/components/UI/Button.tsx]: looks like an indentation issue

File: client/src/components/UI/Button.tsx [hunk 22]
```diff
@@ -0,0 +1,32 @@
+import React from "react";
+import classes from './Button.module.css'
+
+declare interface ButtonProps {
+  children?: string| string[];
+  type?: "submit" | "reset" | "button";
+  classText: string;
```
Comment (13bscsdahmed) [inline client/src/components/UI/Button.tsx]: whats the purpose of class text?

File: client/src/components/UI/Toast.tsx [hunk 23]
```diff
@@ -0,0 +1,58 @@
+import React,{useCallback, useEffect} from "react";
+import { toastActions } from "../../store/toast";
+import { useDispatch } from "react-redux";
+import styles from "./Toast.module.css";
+
+export interface IToastProps {
+    id: number;
```
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: check indentation,. for this file it looks like 4 spaces,. but for other its 2 spaces

File: client/src/components/UI/Toast.tsx [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    },
+    [toastList,dispatch]
+  );
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      if (toastList.length) {
+        deleteToast(toastList[0].id);
+      }
+    }, 3000);
```
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: the timeout should also be configurable,. so the user can also pass the interval for each toast,. if not passed then it should use a default value

File: client/src/components/UI/Toast.tsx [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+    return () => {
+      clearInterval(interval);
+    };
+  }, [toastList, deleteToast]);
+  return (
+    <div className={`${styles.container} ${styles[position]}`}>
+      {toastList.map((toast, i) => (
+        <div
+          key={i}
```
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: use id of toast as key,. we are goign to be adding and deleting also and this will defeat the whole purpose of key

File: client/src/components/UI/Toast.tsx [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+          className={`${styles.notification} ${styles.toast} ${styles[position]}`}
+          style={{ backgroundColor: toast.backgroundColor }}
+        >
+          {toast.description !=='success' ? <button onClick={() => deleteToast(toast.id)}>X</button> : null}
```
Comment (13bscsdahmed) [inline client/src/components/UI/Toast.tsx]: why checking for success,. the closable option can be passed as props and if its true we show the cross

File: client/src/hooks/use-http.ts [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+
+declare interface IConfig{
+    method: string;
+    url:string;
+    data: userData;
+}
+
+
+
+const useHttp = () => {
```
Comment (13bscsdahmed) [inline client/src/hooks/use-http.ts:25]: as discussed offline,. please modify as required

File: client/src/store/toast.ts [hunk 28]
```diff
@@ -0,0 +1,56 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+import { IToastProps } from "../components/UI/Toast";
+
```
Comment (13bscsdahmed) [inline client/src/store/toast.ts:4]: indentation should be consistent

File: client/src/store/toast.ts [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+    toastProp: []
+}
+
+const toastSlice = createSlice({
+    name: 'toast',
+    initialState:  initialToastState,
+    reducers:{
+        showToast(state,action: PayloadAction<IPayload>){
+            switch(action.payload.type) {
+                case 'Success':
```
Comment (13bscsdahmed) [inline client/src/store/toast.ts]: please make an enum for toast types and match against the enum,. hardcoding of string is never a good idea

File: client/src/store/toast.ts [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+                  }
+                  break;
+                default:
+                  break ;
+              }
+            
+            state.toastProp = [...state.toastProp,toastProperties];
+        },
+        removeToast(state,action: PayloadAction<IToastProps[]>){
+            state.toastProp = action.payload;
```
Comment (13bscsdahmed) [inline client/src/store/toast.ts]: does this remove a single toast if so,. we should get the id as payload and remove that particular toast

**Review by 13bscsdahmed — CHANGES_REQUESTED** [review]

File: client/src/store/toast.ts (hunk 30 above)
Comment (FarhadA97) [reply client/src/store/toast.ts]: yes it removes a single toast.. Updated it, now it takes the id as payload.

File: client/src/components/Auth/AuthForm.tsx (hunk 19 above)
Comment (FarhadA97) [reply client/src/components/Auth/AuthForm.tsx]: I looked it up and found that push was required with useHistory..navigate() by default works the same as history.push().

File: client/src/components/UI/Button.tsx (hunk 22 above)
Comment (FarhadA97) [reply client/src/components/UI/Button.tsx]: I'm passing different css classes for different buttons as props. I've changed the name of this prop to className.

File: client/src/components/UI/Toast.tsx (hunk 25 above)
Comment (FarhadA97) [reply client/src/components/UI/Toast.tsx]: updated the key to toast id

File: client/src/pages/AuthPage.tsx [hunk 31]
```diff
@@ -1,9 +1,28 @@
...
-    <AuthForm />
-  )
-}
+  const dispatch = useAppDispatch();
+  const navigate = useNavigate();
+  const loginHandler: (isLogin: boolean, data: User) => void = (isLogin,data) => {
+    let url: string = loginURL;
+    if (!isLogin) url = registerURL;
+    dispatch(login({ data, url }))
+      .unwrap()
```
Comment (13bscsdahmed) [inline client/src/pages/AuthPage.tsx]: why the unwrap?

Comment (13bscsdahmed) [reply client/src/pages/AuthPage.tsx]: i think we just need to dispatch action here and nothing else

File: client/src/store/auth.ts [hunk 32]
```diff
@@ -1,49 +1,91 @@
...
+    builder.addCase(login.pending,(state) => {
+      state.loading = true;
+    })
+    builder.addCase(login.fulfilled, (state, action) => {
+      state.authData = action.payload.user;
+      state.loading = false;
+    });
+    builder.addCase(login.rejected, (state, action) => {
+      state.loading = false;
+      state.errors = action.payload;
```
Comment (13bscsdahmed) [inline client/src/store/auth.ts:87]: nice,. looks very good

**Review by 13bscsdahmed — CHANGES_REQUESTED** [review]

File: client/src/pages/AuthPage.tsx (hunk 31 above)
Comment (FarhadA97) [reply client/src/pages/AuthPage.tsx]: I've fixed this. As discussed, i'm now passing navigate as parameter to dispatch function.
